File: python/src/pmsec/tools/yarn.py
from __future__ import annotations

import re
from pathlib import Path

from pmsec.util.context import Context
from pmsec.util.extras import apply_extras, read_extras, remove_extras
from pmsec.util.io import write_atomic
from pmsec.util.lines import read_key, remove_key, set_key
from pmsec.util.paths import yarnrc_path
from pmsec.util.version import build_preflight

NAME = "yarn"
KEY = "npmMinimalAgeGate"
DOCS = "https://yarnpkg.com/configuration/yarnrc#npmMinimalAgeGate"
MIN_BIN = (4, 10, 0)
SEP = ":"
EXTRAS = [
    {"key": "enableHardenedMode", "expected": "true", "line": "enableHardenedMode: true", "sep": SEP},
    {"key": "enableScripts", "expected": "false", "line": "enableScripts: false", "sep": SEP},
    {"key": "approvedGitRepositories", "expected": "[]", "line": "approvedGitRepositories: []", "sep": SEP},
]

_DURATION = re.compile(r'^"?\s*(\d+)\s*(d|days?|w|weeks?)\s*"?$', re.IGNORECASE)

preflight = build_preflight(
    NAME, MIN_BIN,
    "npmMinimalAgeGate is silently ignored. Upgrade yarn (v4.10+) to enforce the cooldown.",
)


def path(ctx: Context) -> Path:
    return yarnrc_path(ctx.env, ctx.home)


def _parse_days(value: str | None) -> int | None:
    if value is None:
        return None
    m = _DURATION.match(value)
    if not m:
        return None
    n = int(m.group(1))
    return n * 7 if m.group(2).lower() in ("w", "week", "weeks") else n


def read(ctx: Context) -> dict:
    p = path(ctx)
    raw = p.read_text("utf-8") if p.exists() else ""
    value = read_key(raw, KEY, sep=SEP)
    return {"path": str(p), "configured": value, "days": _parse_days(value), "extras": read_extras(raw, EXTRAS)}


def write(days: int, ctx: Context) -> dict:
    p = path(ctx)
    raw = p.read_text("utf-8") if p.exists() else ""
    text = apply_extras(set_key(raw, KEY, f'{KEY}: "{days}d"', sep=SEP), EXTRAS)
    write_atomic(p, text)
    return {"path": str(p)}


def unset(ctx: Context) -> dict:
    p = path(ctx)
    if not p.exists():
        return {"path": str(p), "removed": False}
    before = p.read_text("utf-8")
    after, removed_cooldown = remove_key(before, KEY, sep=SEP)
    after, removed_extras = remove_extras(after, EXTRAS)
    removed = removed_cooldown or removed_extras
    if removed:
        write_atomic(p, after)
    return {"path": str(p), "removed": removed}

File: node/src/tools/yarn.mjs
import { yarnrcPath } from "../util/paths.mjs";
import { readSafe, writeAtomic } from "../util/io.mjs";
import { readKey, setKey, removeKey } from "../util/lines.mjs";
import { readExtras, applyExtras, removeExtras } from "../util/extras.mjs";
import { buildPreflight } from "../util/version.mjs";

export const name = "yarn";
export const key = "npmMinimalAgeGate";
export const docs = "https://yarnpkg.com/configuration/yarnrc#npmMinimalAgeGate";
export const minBin = [4, 10, 0];
const SEP = ":";
export const extras = [
  { key: "enableHardenedMode", expected: "true", line: "enableHardenedMode: true", sep: SEP },
  { key: "enableScripts", expected: "false", line: "enableScripts: false", sep: SEP }
];

export function path(ctx) { return yarnrcPath(ctx.env, ctx.home); }

export const preflight = buildPreflight(name, minBin,
  "npmMinimalAgeGate is silently ignored. Upgrade yarn (v4.10+) to enforce the cooldown.");

function parseDays(value) {
  if (value === null) return null;
  const m = value.match(/^"?\s*(\d+)\s*(d|days?|w|weeks?)\s*"?$/i);
  if (!m) return null;
  const n = Number(m[1]);
  return /^(w|weeks?)$/i.test(m[2]) ? n * 7 : n;
}

export async function read(ctx) {
  const p = path(ctx);
  const raw = await readSafe(p);
  const value = readKey(raw, key, { sep: SEP });
  return {
    path: p, configured: value, days: parseDays(value),
    extras: readExtras(raw, extras)
  };
}

export async function write(days, ctx) {
  const p = path(ctx);
  let text = setKey(await readSafe(p), key, `${key}: "${days}d"`, { sep: SEP });
  text = applyExtras(text, extras);
  await writeAtomic(p, text);
  return { path: p };
}

export async function unset(ctx) {
  const p = path(ctx);
  const before = await readSafe(p);
  const cooldown = removeKey(before, key, { sep: SEP });
  const ex = removeExtras(cooldown.text, extras);
  const removed = cooldown.removed || ex.removed;
  if (removed) await writeAtomic(p, ex.text);
  return { path: p, removed };
}
